Add NewTicker() to Clock

ash2k opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

See linked issues. TL;DR adds a missing method to allow safer usage and to bridge the gap between two copies of the clock package so that the one in apimachinery can be deleted.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

Fixes https://github.com/kubernetes/utils/issues/196.

Updates https://github.com/kubernetes/kubernetes/issues/94738.

**Special notes for your reviewer**:


**Release note**:
```
NONE
```

liggitt commented:
Thanks for the PR. make sure all the new methods in the interface have associate godoc.

Also, since golang has not marked the time methods as deprecated, I don't think we should mark ours as deprecated. Doing so will cause golint to fail.

liggitt commented:
I also suspect adding a method to the interface will get flagged as an API incompatibility in the apidiff presubmit (which is currently failing early because of a github actions change)

apelisse commented:
You can work around the API change by making a "ClockV2" Interface that embeds the other one and adds the method. And a "NewClockV2" that returns that interface (but actually returns the exact same object).

ash2k commented:
@apelisse Thanks for the tip. I've pushed the changes.

apelisse commented on clock/clock.go:
```diff
@@ -30,13 +30,36 @@ type PassiveClock interface {
-var _ = Clock(RealClock{})
+// V2 allows for injecting fake or real clocks into code that
+// needs to do arbitrary things based on time.
+type V2 interface {
```
V2 is not a great name :-)

apelisse replied:
Since the interfaces are pretty much frozen once built, I think you can call it "ClockWithTicker" or something like this.

ash2k replied:
I've tried both `ClockV2` and `ClockWithTicker` but linted doesn't like such names - see https://github.com/kubernetes/utils/pull/197/checks?check_run_id=1685665492#step:4:62 That's why I tried `V2`, which you don't like :)

apelisse replied:
Yeah, "WithTicker" is fine (to me at least ;-))

ash2k replied:
@apelisse PTAL

ash2k commented:
/lifecycle frozen

ash2k commented:
/remove-lifecycle stale

ash2k commented:
@apelisse @liggitt @lavalamp Would be awesome to get this merged (I think I've addressed all comments) so that I can start working on https://github.com/kubernetes/kubernetes/issues/94738.

lavalamp commented on clock/clock.go:
```diff
@@ -30,13 +30,36 @@ type PassiveClock interface {
 	Sleep(d time.Duration)
+	// Tick returns the channel of a new Ticker.
+	// This method does not allow to free/GC the backing ticker. Use
+	// NewTicker instead.
```
Explain: how does one call NewTicker? It's not in this interface?

ash2k replied:
I've clarified the comment :)

lavalamp commented on clock/clock.go:
```diff
@@ -30,13 +30,36 @@ type PassiveClock interface {
-var _ = Clock(RealClock{})
+// WithTicker allows for injecting fake or real clocks into code that
+// needs to do arbitrary things based on time.
+type WithTicker interface {
```
I think it'd be much more convenient if there were either a) a ClockV2 interface, or b) we made a v2 version of the clock package. (and v1 can be a wrapper that just calls the v2 package.) The latter is probably most correct but it might take some mucking about with go mod files.

ash2k replied:
I've tried `ClockV2` - https://github.com/kubernetes/utils/pull/197#discussion_r555531160

I'm not sure how to make a "v2 version of the clock package". I can make a v2 of the whole repo, but I'm not sure what you mean here, sorry.

ash2k replied:
I think it might be worth considering a `v2` of the package rather than adding a new clock interface. A new major version shouldn't break anyone who is using go modules.

lavalamp replied:
It's unfortunate if we have to go to v2 of everything in utils. @liggitt might know if we can do some go mod magic to make clock its own module without breaking anyone, and then bump just that to v2. I think it might be possible?

@thockin might have an opinion on whether it's worth rev'ing everything under utils if it's hard to just rev the clock package.

liggitt replied:
> @liggitt might know if we can do some go mod magic to make clock its own module without breaking anyone, and then bump just that to v2. I think it might be possible?

There's no magic. It's possible but really tedious to manage multiple published modules in a single repo (releases have to have per-module tags, etc). I think you'd get about the same effect with a `k8s.io/utils/clock/v2` package

lavalamp replied:
Yeah that actually sounds perfect, if a little bit of work to set up (e.g. we'd want the regular clock package to just call into the v2 package for as much as possible).

MikeSpreitzer commented:
/test continuous-integration/travis-ci

apelisse commented on clock/testing/fake_clock.go:
```diff
@@ -231,6 +249,12 @@ func (*IntervalClock) Tick(d time.Duration) <-chan time.Time {
 
+// NewTicker is currently unimplemented, will panic.
+// TODO: make interval clock use FakeClock so this can be implemented.
+func (*IntervalClock) NewTicker(d time.Duration) clock.Ticker {
```
Why do we need this function again? If we didn't, IntervalClock wouldn't implement the "WithTicker" interface, which means that it would fail at compile time. By adding it, it now implements WithTicker, but will fail at runtime?

MikeSpreitzer replied:
True.  BTW, IntervalClock also has a lot of unimplemented methods for `Clock`.

MikeSpreitzer replied:
We could remove them all, leaving IntervalClock implementing `PassiveClock` --- which would be a compile-time tight characterization.

MikeSpreitzer replied:
There are no uses of IntervalClock in k8s.io/utils.

In the main repo there are only two files that use it: `client-go/tools/record/event_test.go` and `client-go/tools/record/events_cache_test.go`.

lavalamp replied:
Let's make it possible to remove the api machinery clock. We should also not implement interfaces that don't actually work -- this library is for public consumption and that will be very annoying to a consumer.

MikeSpreitzer replied:
Ugh, there is non-trivial entanglement with code that needs a non-passive clock.  In `client-go/tools/record/events_cache.go`, the Filter method of EventSourceObjectSpamFilter passes its clock to `flowcontrol.NewTokenBucketRateLimiterWithClock` --- which needs a clock that can `Sleep`.  Why does this work?

MikeSpreitzer replied:
@lavalamp : I think #212 does what you said.

ash2k commented:
Closing in favor of https://github.com/kubernetes/utils/pull/212.
